### PR TITLE
Add a first-class way of accessing caller's exports

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is a continuation of #1237 and updates the API of `Func` to
allow defining host functions which have easy access to a caller's
memory in particular. The new APIs look like so:

* The `Func::wrap*` family of functions was condensed into one
  `Func::wrap` function.
* The ABI layer of conversions in `WasmTy` were removed
* An optional `Caller<'_>` argument can be at the front of all
  host-defined functions now.

The old way the wasi bindings looked up memory has been removed and is
now replaced with the `Caller` type. The `Caller` type has a
`get_export` method on it which allows looking up a caller's export by
name, allowing you to get access to the caller's memory easily, and even
during instantiation.

### CODE REVIEW

**Comment by fitzgen** [comment]
Gah it looks like the subscribe-to-label action also needs to run on a cron schedule for pull requests: https://github.com/bytecodealliance/wasmtime/pull/1290/checks?check_run_id=501503929#step:2:25

File: crates/api/src/func.rs [hunk 1]
```diff
@@ -995,3 +829,194 @@ impl<T: WasmTy> WasmRet for Result<T, Trap> {
...
+    #[doc(hidden)]
+    fn into_func(self, store: &Store) -> Func;
+}
+
+/// A structure representing the *caller's* context when creating a function
+/// via [`Func::wrap`].
+///
+/// This structure can be taken as the first parameter of a closure passed to
+/// [`Func::wrap`], and it can be used to learn information about the caller of
+/// the function, such as the calling module's memory, exports, etc.
```
Comment (sunfishcode) [inline crates/api/src/func.rs]: It'd be good to mention somewhere that this is intended to be a temporary mechanism, and that we're hoping interface types will provide better ways to do this in the future.

File: crates/api/src/func.rs [hunk 2]
```diff
@@ -995,3 +829,194 @@ impl<T: WasmTy> WasmRet for Result<T, Trap> {
...
+    caller_vmctx: *mut VMContext,
+}
+
+impl Caller<'_> {
+    /// Looks up an export from the caller's module by the `name` given.
+    ///
+    /// The primary purpose of this structure is to provide access to the
+    /// caller's information, such as it's exported memory. This allows
+    /// functions which take pointers as arguments to easily read the memory the
+    /// pointers point into.
```
Comment (sunfishcode) [inline crates/api/src/func.rs]: It feels like this paragraph belongs on the `struct` rather than on the `get_export` method.

**Comment by alexcrichton** [comment]
Ok, updated!

**Pull request merged**